### ISSUE TITLE
Canvas: trying to draw too large(143769912bytes)

### ISSUE THREAD

**Issue by logic** [stefan-niedermann/nextcloud-notes#1034]
Opening up a markdown note with a particularly long checklist in it (in preview mode; edit mode is fine) with the latest fdroid release throws this error after you scroll down a little bit (the initial page of rendering is fine):

```
App Version: 2.19.0
App Version Code: 2019000
App Flavor: fdroid

Files App Version Code: 30140190

---

OS Version: 3.18.91-g8bbffc1d427(eng.root.20210107.120301)
OS API Level: 29
Device: heroltexx
Manufacturer: samsung
Model (and Product): SM-G930F (heroltexx)

---

java.lang.RuntimeException: Canvas: trying to draw too large(143769912bytes) bitmap.
	at android.graphics.RecordingCanvas.throwIfCannotDraw(RecordingCanvas.java:280)
	at android.graphics.BaseRecordingCanvas.drawBitmap(BaseRecordingCanvas.java:88)
	at android.graphics.drawable.BitmapDrawable.draw(BitmapDrawable.java:548)
	at io.noties.markwon.image.AsyncDrawable.draw(AsyncDrawable.java:326)
	at io.noties.markwon.image.AsyncDrawableSpan.draw(AsyncDrawableSpan.java:126)
	at android.text.TextLine.handleReplacement(TextLine.java:1011)
	at android.text.TextLine.handleRun(TextLine.java:1158)
	at android.text.TextLine.drawRun(TextLine.java:491)
	at android.text.TextLine.draw(TextLine.java:286)
	at android.text.Layout.drawText(Layout.java:576)
	at android.widget.Editor.drawHardwareAcceleratedInner(Editor.java:1957)
	at android.widget.Editor.drawHardwareAccelerated(Editor.java:1876)
	at android.widget.Editor.onDraw(Editor.java:1816)
	at android.widget.TextView.onDraw(TextView.java:7989)
	at android.view.View.draw(View.java:21975)
	at android.view.View.updateDisplayListIfDirty(View.java:20852)
	at android.view.View.draw(View.java:21707)
	at android.view.ViewGroup.drawChild(ViewGroup.java:4432)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4193)
	at android.view.View.draw(View.java:21978)
	at android.widget.ScrollView.draw(ScrollView.java:1835)
	at android.view.View.updateDisplayListIfDirty(View.java:20852)
	at android.view.View.draw(View.java:21707)
	at android.view.ViewGroup.drawChild(ViewGroup.java:4432)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4193)
	at android.view.View.draw(View.java:21978)
	at android.view.View.updateDisplayListIfDirty(View.java:20852)
	at android.view.View.draw(View.java:21707)
	at android.view.ViewGroup.drawChild(ViewGroup.java:4432)
	at androidx.coordinatorlayout.widget.CoordinatorLayout.drawChild(CoordinatorLayout.java:1277)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4193)
	at android.view.View.updateDisplayListIfDirty(View.java:20843)
	at android.view.View.draw(View.java:21707)
	at android.view.ViewGroup.drawChild(ViewGroup.java:4432)
	at androidx.fragment.app.FragmentContainerView.drawChild(FragmentContainerView.java:235)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4193)
	at androidx.fragment.app.FragmentContainerView.dispatchDraw(FragmentContainerView.java:223)
	at android.view.View.draw(View.java:21978)
	at android.view.View.updateDisplayListIfDirty(View.java:20852)
	at android.view.View.draw(View.java:21707)
	at android.view.ViewGroup.drawChild(ViewGroup.java:4432)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4193)
	at android.view.View.updateDisplayListIfDirty(View.java:20843)
	at android.view.View.draw(View.java:21707)
	at android.view.ViewGroup.drawChild(ViewGroup.java:4432)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4193)
	at android.view.View.updateDisplayListIfDirty(View.java:20843)
	at android.view.View.draw(View.java:21707)
	at android.view.ViewGroup.drawChild(ViewGroup.java:4432)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4193)
	at android.view.View.updateDisplayListIfDirty(View.java:20843)
	at android.view.View.draw(View.java:21707)
	at android.view.ViewGroup.drawChild(ViewGroup.java:4432)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4193)
	at android.view.View.updateDisplayListIfDirty(View.java:20843)
	at android.view.View.draw(View.java:21707)
	at android.view.ViewGroup.drawChild(ViewGroup.java:4432)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4193)
	at android.view.View.updateDisplayListIfDirty(View.java:20843)
	at android.view.View.draw(View.java:21707)
	at android.view.ViewGroup.drawChild(ViewGroup.java:4432)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4193)
	at android.view.View.draw(View.java:21978)
	at com.android.internal.policy.DecorView.draw(DecorView.java:808)
	at android.view.View.updateDisplayListIfDirty(View.java:20852)
	at android.view.ThreadedRenderer.updateViewTreeDisplayList(ThreadedRenderer.java:581)
	at android.view.ThreadedRenderer.updateRootDisplayList(ThreadedRenderer.java:587)
	at android.view.ThreadedRenderer.draw(ThreadedRenderer.java:664)
	at android.view.ViewRootImpl.draw(ViewRootImpl.java:3767)
	at android.view.ViewRootImpl.performDraw(ViewRootImpl.java:3495)
	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2779)
	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1745)
	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:7768)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:967)
	at android.view.Choreographer.doCallbacks(Choreographer.java:791)
	at android.view.Choreographer.doFrame(Choreographer.java:726)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:952)
	at android.os.Handler.handleCallback(Handler.java:883)
	at android.os.Handler.dispatchMessage(Handler.java:100)
	at android.os.Looper.loop(Looper.java:214)
	at android.app.ActivityThread.main(ActivityThread.java:7356)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:491)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:940)

```

**Comment by stefan-niedermann** [comment]
- Can you please share a minimal sample markdown (just attach here as a text file or at pastebin or whatever) to reproduce this issue?
- Do you have any image links (`![my image](https://example.com/image.jpg)`) inside of this note or only task lists?

**Comment by logic** [comment]
Ah! Yes, there are a couple of image links at the end of the document. And yeah, they (well, with a bit of testing, one particularly large image) seem to be the problem, not the length of the task list.

**Comment by stefan-niedermann** [comment]
Can you please share the link to this image, so i can investigate this issue further? (or send it via mail ton info@niedermann.it if it ain't public)

**Comment by logic** [comment]
One image causing this consistently is: http://otakurevolution.com/storyimgs/falldog/GundamTimeline/Falldogs_GundamTimeline_v13_April2020.png

I'll see if I can find any other examples I can share.

**Comment by stefan-niedermann** [comment]
[I have reported this issue upstream](https://github.com/noties/Markwon/issues/329) in the library we are using because i currently can't see a way how to catch this properly in the Notes app.

**Comment by stefan-niedermann** [comment]
Update: Upstream found a solution which I already was able to test & verify:

It will take some time until upstream released a new version and we can publish a new Notes version with the fix, but it's just a matter of time now.

[![grafik](https://user-images.githubusercontent.com/4741199/105827722-9f8c5380-5fc2-11eb-845c-7853cace7f4a.png)](https://user-images.githubusercontent.com/4741199/105826820-76b78e80-5fc1-11eb-85f9-632aef5b0f4c.png)

**Comment by stefan-niedermann** [comment]
Will be fixed in `v3.1.2`
